nasbench/decoder.py

```diff
@@ -73,25 +75,40 @@ def __init__(self,
         for i in range(self.n):
             self.offsets.append( (i + 3) * i // 2 - 1)
     
-    def forward(self, x, encoder_hidden=None, encoder_outputs=None):
+    def forward(self, x, x_len, encoder_hidden=None, encoder_outputs=None):
+        # x is decoder_inputs = [0] + encoder_inputs[:-1]
+
         decoder_hidden = self._init_state(encoder_hidden)
         if x is not None:
             bsz = x.size(0)
             tgt_len = x.size(1)
             x = self.embedding(x)
             x = F.dropout(x, self.dropout, training=self.training)
             residual = x
+            
+            x = pack_padded_sequence(x, x_len, batch_first=True)
             x, hidden = self.rnn(x, decoder_hidden)
+            x = pad_packed_sequence(x, batch_first=True)[0]
+
             x = (residual + x) * math.sqrt(0.5)
             residual = x
-            x, _ = self.attention(x, encoder_outputs)
+            
+            # create mask
+            mask = torch.zeros(bsz, x.size(1))
+            for i,l in enumerate(x_len):
+                for j in range(l):
+                    mask[i][j] = 1
+            mask = (mask == 0).unsqueeze(1)
+            mask = utils.move_to_cuda(mask)
+            
+            x, _ = self.attention(x, encoder_outputs, mask=mask)
             x = (residual + x) * math.sqrt(0.5)
             predicted_softmax = F.log_softmax(self.out(x.view(-1, self.hidden_size)), dim=-1)
             predicted_softmax = predicted_softmax.view(bsz, tgt_len, -1)
             return predicted_softmax, None
```

Comment:
Is predicted softmax returns sane values? If padded elements are zero-initialized, then probabilty values will be broken.